docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

```diff
@@ -0,0 +1,1951 @@
+    pure alice
+  
+  getIOU : Party -> Script (Optional (ContractId IOU, IOU))
+  getIOU key = queryContractKey @IOU key key
+
+All other files should remain the same.
+
+Changing Choices
+~~~~~~~~~~~~~~~~
+
+Add the following choice, ``Duplicate``, to both v1 and v2 versions of IOU:
+
+.. code:: daml
+
+      choice Duplicate : ContractId IOU
+        controller issuer
+        do
+          create this with value = value * 2
+
+Running ``daml build`` should succeed without errors.
+
+.. code:: bash
+
+  > cd v1/my-pkg
+  > daml build
+  ...
+  Created .daml/dist/my-pkg-1.0.0.dar
+  > cd ../../v2/my-pkg
+  > daml build
+  ...
+  Created .daml/dist/my-pkg-1.1.0.dar
+
+We can upgrade the ``Duplicate`` choice by adding an optional field ``amount``,
+and changing the behavior of the choice to default to a multiple of 3.
+Replace the definition of the ``Duplicate`` choice in the v2 implementation
+of ``IOU`` only:
+
+.. code:: daml
+
+  ...
+  -- Add import to top of module
+  import DA.Optional (fromOptional)
+  ...
+     -- Replace Duplicate choice implementation
+     choice Duplicate : ContractId IOU
+       with
+         amount : Optional Int -- New optional amount
+       controller issuer
+       do
+         let amt = fromOptional 3 amount
+         create this with value = value * amt
+  ...
+
+Add a script called ``duplicateIOU`` in V1:
+
+.. code:: daml
+
+  ...
+  duplicateIOU : Party -> Script (Optional (ContractId IOU, IOU))
+  duplicateIOU key = do
+    mbIOU <- getIOU key
+    case mbIOU of
+      None -> pure None
+      Some (iouCid, _) -> do
+        newCid <- key `submit` exerciseExactCmd iouCid Duplicate
+        mbNewIOU <- queryContractId key newCid
+        case mbNewIOU of
+          Some newIOU -> pure (Some (newCid, newIOU))
+          None -> pure None
+
+A similar script called ``duplicateIOU`` should be added in V2, supplying an
+``amount`` field:
+
+.. code:: daml
+
+  ...
+  duplicateIOU : Party -> Script (Optional (ContractId IOU, IOU))
+  duplicateIOU key = do
+    mbIOU <- getIOU key
+    case mbIOU of
+      None -> pure None
+      Some (iouCid, _) -> do
+        newCid <- key `submit` exerciseExactCmd iouCid Duplicate { amount = Some 4 }
+        mbNewIOU <- queryContractId key newCid
+        case mbNewIOU of
+          Some newIOU -> pure (Some (newCid, newIOU))
+          None -> pure None
+
+Running the v1 ``duplicateIOU`` script with ``exerciseExactCmd`` always runs
+the v1 implementation for the ``Duplicate`` choice, and likewise for v2.
+
+Modifying Signatory Definitions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Other definitions can be changed, but warnings are emitted to remind the
+developer that the changes can be unsafe and need to be made with care
+to preserve necessary invariants.
+
+Signatories and observers are one expression that can be changed.
+Importantly, the new definition must not change the resulting list of
+signatories and observers for existing contracts.
```

Comment:
```suggestion
Importantly, the SCU assumes that the new definition does not alter the computation of the signatories, the observers, and the key for existing contracts. Otherwise, dynamic contract upgrades will fail at runtime.
```